### PR TITLE
Use AP::airspeed() to get airspeed sensor, rather than going via AP_AHRS

### PR DESCRIPTION
**Pull request by peterbarker**
Setting the pointer in the AHRS predates our use of singletons.

Some of the users of this should probably be failling back to using a synthetic airspeed estimate instead of giving up if we don't have a true airspeed sensor.

### CODE REVIEW

File: Tools/Replay/Replay.cpp [hunk 1]
```diff
@@ -654,7 +654,6 @@ void Replay::read_sensors(const char *type)
     } else if (streq(type,"MAG")) {
         _vehicle.compass.read();
     } else if (streq(type,"ARSP")) {
-        _vehicle.ahrs.set_airspeed(&_vehicle.airspeed);
```
Comment (magicrub) [inline Tools/Replay/Replay.cpp:657]: Is this now an empty else if?

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp [hunk 2]
```diff
@@ -697,7 +697,7 @@ AP_AHRS_DCM::drift_correction(float deltat)
         }
         float airspeed;
         if (airspeed_sensor_enabled()) {
-            airspeed = _airspeed->get_airspeed();
+            airspeed = AP::airspeed()->get_airspeed();
```
Comment (magicrub) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp:700]: Needs a null check. Not all vehicles instantiate airspeed

Comment (peterbarker) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp:700]: airspeed_sensor_enabled() does that.  Further, we're already dereferencing a pointer :-)

I do want to rename `airspeed_sensor_enabled()` to `using_airspeed_sensor()`.

File: Tools/Replay/Replay.cpp (hunk 1 above)
Comment (peterbarker) [reply Tools/Replay/Replay.cpp:657]: It is :-)

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp (hunk 2 above)
Comment (magicrub) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp:701]: I meant AP_::airspeed() could be nullptr

Comment (peterbarker) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp:701]: No, it can't; `airspeed_sensor_enabled` checks for nullptr.  It already checks `_airspeed` for nullptr, and has been changed to check the singleton instead.

File: Tools/Replay/Replay.cpp (hunk 1 above)
Comment (peterbarker) [reply Tools/Replay/Replay.cpp]: .... and now it no longer exists :-)

File: libraries/AP_AHRS/AP_AHRS.h [hunk 3]
```diff
@@ -285,12 +277,14 @@ class AP_AHRS
     // return true if airspeed comes from an airspeed sensor, as
     // opposed to an IMU estimate
     bool airspeed_sensor_enabled(void) const {
+        const AP_Airspeed     * _airspeed = AP::airspeed();
```
Comment (magicrub) [inline libraries/AP_AHRS/AP_AHRS.h:280]: unnecessary whitespace

File: libraries/AP_AHRS/AP_AHRS.h [hunk 4]
```diff
@@ -285,12 +277,14 @@ class AP_AHRS
...
         return _airspeed != nullptr && _airspeed->use() && _airspeed->healthy();
     }
 
     // return true if airspeed comes from a specific airspeed sensor, as
     // opposed to an IMU estimate
     bool airspeed_sensor_enabled(uint8_t airspeed_index) const {
+        const AP_Airspeed     * _airspeed = AP::airspeed();
```
Comment (magicrub) [inline libraries/AP_AHRS/AP_AHRS.h:287]: unnecessary whitespace

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp (hunk 2 above)
Comment (magicrub) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp:701]: This is the same argument that was discussed on yesterday's dev call. The function should return the result when true, in this case a non-null ptr. Having a function check something, and then at a later time it is assumed true. That's not thread safe.

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp [hunk 5]
```diff
@@ -1049,14 +1049,14 @@ bool AP_AHRS_DCM::get_position(struct Location &loc) const
 // return an airspeed estimate if available
 bool AP_AHRS_DCM::airspeed_estimate(float &airspeed_ret) const
 {
-    return airspeed_estimate(_airspeed?_airspeed->get_primary():0, airspeed_ret);
+    return airspeed_estimate(AP::airspeed()?AP::airspeed()->get_primary():0, airspeed_ret);
```
Comment (magicrub) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp:1052]: should not be checking if non-zero, should explicitly looking for nullptr

Comment (magicrub) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp:1052]:
```suggestion
    return airspeed_estimate((AP::airspeed() != nullptr)?AP::airspeed()->get_primary():0, airspeed_ret);
```

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp [hunk 6]
```diff
@@ -1049,14 +1049,14 @@ bool AP_AHRS_DCM::get_position(struct Location &loc) const
...
 }
 
 // return an airspeed estimate from a specific airspeed sensor instance if available
 bool AP_AHRS_DCM::airspeed_estimate(uint8_t airspeed_index, float &airspeed_ret) const
 {
     if (airspeed_sensor_enabled(airspeed_index)) {
-        airspeed_ret = _airspeed->get_airspeed(airspeed_index);
+        airspeed_ret = AP::airspeed()->get_airspeed(airspeed_index);
```
Comment (magicrub) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp:1059]: AP::airspeed() needs nullptr check

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp [hunk 7]
```diff
@@ -2231,6 +2231,7 @@ uint8_t AP_AHRS_NavEKF::get_active_airspeed_index() const
     }
 #endif
     // for the rest, let the primary airspeed sensor be used
+    const AP_Airspeed     * _airspeed = AP::airspeed();
```
Comment (magicrub) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp:2234]: unnecessary whitespace

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp [hunk 8]
```diff
@@ -516,7 +516,7 @@ uint32_t AP_Frsky_SPort_Passthrough::calc_velandyaw(void)
     AP_AHRS &_ahrs = AP::ahrs();
     WITH_SEMAPHORE(_ahrs.get_semaphore());
     // horizontal velocity in dm/s (use airspeed if available and enabled - even if not used - otherwise use groundspeed)
-    const AP_Airspeed *aspeed = _ahrs.get_airspeed();
+    const AP_Airspeed *aspeed = AP::airspeed();
     if (aspeed && aspeed->enabled()) {
```
Comment (magicrub) [inline libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:520]: needs explicit nullptr check instead of non-null

File: libraries/AP_LTM_Telem/AP_LTM_Telem.cpp [hunk 9]
```diff
@@ -121,7 +121,7 @@ void AP_LTM_Telem::send_Sframe(void)
     const uint16_t amp = (uint16_t) roundf(current * 100.0f);                          // current sensor (expects value in hundredth of A)
 
     // airspeed in m/s if available and enabled - even if not used - otherwise send 0
-    const AP_Airspeed *aspeed = AP::ahrs().get_airspeed();
+    const AP_Airspeed *aspeed = AP::airspeed();
     uint8_t airspeed = 0; // airspeed sensor (m/s)
     if (aspeed && aspeed->enabled()) {
```
Comment (magicrub) [inline libraries/AP_LTM_Telem/AP_LTM_Telem.cpp:126]: needs explicit nullptr chk instead of nonzero

File: libraries/AP_RCTelemetry/AP_Spektrum_Telem.cpp [hunk 10]
```diff
@@ -435,7 +435,7 @@ void AP_Spektrum_Telem::calc_airspeed()
     AP_AHRS &ahrs = AP::ahrs();
     WITH_SEMAPHORE(ahrs.get_semaphore());
 
-    const AP_Airspeed *airspeed = ahrs.get_airspeed();
+    const AP_Airspeed *airspeed = AP::airspeed();
     float speed = 0.0f;
     if (airspeed && airspeed->healthy()) {
```
Comment (magicrub) [inline libraries/AP_RCTelemetry/AP_Spektrum_Telem.cpp:440]: needs explicit nullptr chk instead of nonzero

**Review by magicrub — CHANGES_REQUESTED** [review]
needs nullptr checking

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp (hunk 5 above)
Comment (tridge) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp:1052]: avoid the double call to singleton:
  auto *airspeed = xxxxx;

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp (hunk 5 above)
Comment (peterbarker) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp]: Fixed.

Comment (peterbarker) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp]: Fixed.

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp (hunk 6 above)
Comment (peterbarker) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp:1067]: This is checked in `airspeed_sensor_enabled`.

I'm looking at refactoring a whole bunch of this stuff and would prefer to not make too many changes.  Note that `_airspeed` here is *already* a pointer, so the nullptr check better work already!

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp (hunk 7 above)
Comment (peterbarker) [reply libraries/AP_AHRS/AP_AHRS_NavEKF.cpp]: Fixed.

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp (hunk 8 above)
Comment (peterbarker) [reply libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:520]: It's modifying a line which isn't modified already in this PR.

I've just changed the way the pointer is fetched.  This pointer will never be nullptr more often that the existing code!

Similarly in the other telem backends.

Those are all probably wrong anyway - they should be asking the AHRS for the estimate, not the sensor.

File: libraries/AP_LTM_Telem/AP_LTM_Telem.cpp (hunk 9 above)
Comment (peterbarker) [reply libraries/AP_LTM_Telem/AP_LTM_Telem.cpp:126]: As elsewhere, unrelated to my PR.  Not a bad change, just unrelated.

File: libraries/AP_RCTelemetry/AP_Spektrum_Telem.cpp (hunk 10 above)
Comment (peterbarker) [reply libraries/AP_RCTelemetry/AP_Spektrum_Telem.cpp:440]: As elsewhere, unrelated to my PR.  Not a bad change, just unrelated.

**Review by tridge — APPROVED** [review]

**Review by magicrub — APPROVED** [review]
